### ISSUE TITLE
Scoping bug of %^H?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2884]
Migrated from [rt.perl.org#4746](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4746) (status was 'resolved')

Searchable as RT4746$

**Comment by p5pRT** [comment]
### From @andk

#### Created by k@ak-71.mind.de


Today I was interested in finding out how one is supposed to
use charnames\.pm at runtime but then I stumbled over this example\.

The following program works as expected and prints the same character
tree times in a row\. I omit the output because it would be converted
to death anyway\. Note that $F\[1\] contains the string
&nbsp;    OGHAM SPACE MARK
after the split\(\)\.

% /usr/local/perl\-5\.7\.0@&#8203;7787/bin/perl \-e '
use charnames qw\(&#8203;:full\); 
@&#8203;ARGV = grep \-e\, map "$\_/unicode/Unicode\.301"\, @&#8203;INC;
LINE&#8203;: while \(\<>\)\{
 next unless /OGHAM/;
 chomp\($\_\);
 @&#8203;F = split /;/\, $\_\, 0;
 use charnames qw\(&#8203;:full\);
 my $evaled = eval\(qq\{"\\\\N\{$F\[1\]\}"\}\);
 my $runtime = charnames&#8203;::charnames\($F\[1\]\);
 printf "%s %s %s %s %s\\n"\, chr\(hex\($F\[0\]\)\)\, $evaled\, $runtime\, @&#8203;F\[0\,1\];
 last;
\}
'

But if I comment out the first occurrence of 'use charnames'\, then
the program does not want to live&#8203;:

% /usr/local/perl\-5\.7\.0@&#8203;7787/bin/perl \-e '
\# use charnames qw\(&#8203;:full\);
@&#8203;ARGV = grep \-e\, map "$\_/unicode/Unicode\.301"\, @&#8203;INC;
LINE&#8203;: while \(\<>\)\{
 next unless /OGHAM/;
 chomp\($\_\);
 @&#8203;F = split /;/\, $\_\, 0;
 use charnames qw\(&#8203;:full\);
 my $evaled = eval\(qq\{"\\\\N\{$F\[1\]\}"\}\);
 my $runtime = charnames&#8203;::charnames\($F\[1\]\);
 printf "%s %s %s %s %s\\n"\, chr\(hex\($F\[0\]\)\)\, $evaled\, $runtime\, @&#8203;F\[0\,1\];
 last;
\}
'
Unknown charname 'OGHAM SPACE MARK' at /usr/local/perl\-5\.7\.0@&#8203;7787/lib/5\.7\.0/charnames\.pm line 32\, \<> line 4249\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.7.0:

Configured by k at Tue Nov 21 02:41:49 CET 2000.

Summary of my perl5 (revision 5.0 version 7 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.16a, archname=i686-linux
    uname='linux ak-71.mind.de 2.2.16a #1 smp wed jun 14 11:59:53 cest 2000 i686 unknown '
    config_args='-Dprefix=/usr/local/perl-5.7.0@7787 -Doptimize=-g -des -Duseperlio -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL7776


@INC for perl v5.7.0:
    /usr/local/perl-5.7.0@7787/lib/5.7.0/i686-linux
    /usr/local/perl-5.7.0@7787/lib/5.7.0
    /usr/local/perl-5.7.0@7787/lib/site_perl/5.7.0/i686-linux
    /usr/local/perl-5.7.0@7787/lib/site_perl/5.7.0
    /usr/local/perl-5.7.0@7787/lib/site_perl
    .


Environment for perl v5.7.0:
    HOME=/home/k
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/k/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/local/perl/bin:/usr/X11/bin:/sbin:/usr/sbin:/usr/local/Office51/bin
    PERL_BADLANG (unset)
    SHELL=/bin/zsh


```
</details>

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Wed\, 22 Nov 2000 21&#8203;:56&#8203;:58 \-0500\, Ilya Zakharevich \<ilya@&#8203;math\.ohio\-state\.edu> said&#8203;:

&nbsp;  > \+\+\+ \./lib/charnames\.pm    Wed Nov 22 21&#8203;:31&#8203;:37 2000

Ilya\, while you are at charnames\, may I bring the bugreport to your
attention that I filed a few days ago about %^H? Is this a scoping bug
or are my expectations wrong? In other words\, shouldn't both programs
behave the same way? Here it is&#8203;:

&nbsp;  The following program works as expected and prints the same character
&nbsp;  tree times in a row\. I omit the output because it would be converted
&nbsp;  to death anyway\. Note that $F\[1\] contains the string
&nbsp;      OGHAM SPACE MARK
&nbsp;  after the split\(\)\.

&nbsp;  % /usr/local/perl\-5\.7\.0@&#8203;7787/bin/perl \-e '
&nbsp;  use charnames qw\(&#8203;:full\); 
&nbsp;  @&#8203;ARGV = grep \-e\, map "$\_/unicode/Unicode\.301"\, @&#8203;INC;
&nbsp;  LINE&#8203;: while \(\<>\)\{
&nbsp;   next unless /OGHAM/;
&nbsp;   chomp\($\_\);
&nbsp;   @&#8203;F = split /;/\, $\_\, 0;
&nbsp;   use charnames qw\(&#8203;:full\);
&nbsp;   my $evaled = eval\(qq\{"\\\\N\{$F\[1\]\}"\}\);
&nbsp;   my $runtime = charnames&#8203;::charnames\($F\[1\]\);
&nbsp;   printf "%s %s %s %s %s\\n"\, chr\(hex\($F\[0\]\)\)\, $evaled\, $runtime\, @&#8203;F\[0\,1\];
&nbsp;   last;
&nbsp;  \}
&nbsp;  '

&nbsp;  But if I comment out the first occurrence of 'use charnames'\, then
&nbsp;  the program does not want to live&#8203;:

&nbsp;  % /usr/local/perl\-5\.7\.0@&#8203;7787/bin/perl \-e '
&nbsp;  \# use charnames qw\(&#8203;:full\);
&nbsp;  @&#8203;ARGV = grep \-e\, map "$\_/unicode/Unicode\.301"\, @&#8203;INC;
&nbsp;  LINE&#8203;: while \(\<>\)\{
&nbsp;   next unless /OGHAM/;
&nbsp;   chomp\($\_\);
&nbsp;   @&#8203;F = split /;/\, $\_\, 0;
&nbsp;   use charnames qw\(&#8203;:full\);
&nbsp;   my $evaled = eval\(qq\{"\\\\N\{$F\[1\]\}"\}\);
&nbsp;   my $runtime = charnames&#8203;::charnames\($F\[1\]\);
&nbsp;   printf "%s %s %s %s %s\\n"\, chr\(hex\($F\[0\]\)\)\, $evaled\, $runtime\, @&#8203;F\[0\,1\];
&nbsp;   last;
&nbsp;  \}
&nbsp;  '
&nbsp;  Unknown charname 'OGHAM SPACE MARK' at /usr/local/perl\-5\.7\.0@&#8203;7787/lib/5\.7\.0/charnames\.pm line 32\, \<> line 4249\.


Thanks\!

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu\, Nov 23\, 2000 at 05&#8203;:50&#8203;:54AM \+0100\, Andreas J\. Koenig wrote&#8203;:
> Ilya\, while you are at charnames\, may I bring the bugreport to your
> attention that I filed a few days ago about %^H? Is this a scoping bug
> or are my expectations wrong? In other words\, shouldn't both programs
> behave the same way? Here it is&#8203;:

Thanks\, it is good that you wrote this; I was thinking whether I want
to beat this issue to death again\.\.\.  Now when the problem report
comes not from me\, I may play the other role\!

>    my $evaled = eval\(qq\{"\\\\N\{$F\[1\]\}"\}\);

%^H \(and I'm sure a lot of interesting flags in $^H\) do not propagate
inside eval ""\.

What is needed to be fixed&#8203;: upgrade pp\_entereval to an SVOP\, and save
%^H in the SV slot at compile\-time \- if %^H is defined when compiling
pp\_entereval\.

Ilya
